### ISSUE TITLE
Using custom Http component

### ISSUE THREAD

**Issue by bozhidarc** [troyanskiy/ng2-resource-rest#19]
I have my own Http class with name AppHttp which extends from Http.
I found in your code that add2Provides property controls if my resource class is added to the app providers with an instance of the original Http class, so I set it to false, but here comes the problem.
When it is false and In the mine implementation of the Resource class i override its constructor and inject  mine Http implementation AppHttp is called twice. The first time it is called with proper backend and defaultOptions arguments but the second time is called with empty backend and defaultOptions which results in the following error `TypeError: backend.createConnection is not a function`.

What is the right way to use the ng2-resource-rest with mine custom Http implemtation?

```
constructor(protected http: AppHttp) {
        super(http);
     }

```

**Comment by bozhidarc** [comment]
I found a solution! In my resource class, I inject the standard Http class after that I casting it to mine AppHttp and then I pass it to the parent constructor.

**Comment by troyanskiy** [comment]
Could you please provide a code of your solution?
It might be useful for me and maybe for other.
Thank you!

**Comment by bozhidarc** [comment]
Here is my code. 
**AppHttp** is created by me and extends the angular 2 Http class.

```
export class PinsResource extends Resource {
    constructor(protected http: Http) {
        super(<AppHttp>http);
     }
}
```

**Comment by radumardale** [comment]
Hi @bozhidarc!
I am trying to do the same thing but it does not seem to work.
I have created a SecuredResource that extends Resource. When using it, the constructor gets called but the methods in my AppHttp don't get called.

What am I missing? 
Thank you.

**Comment by troyanskiy** [comment]
Hi @radumardale
It's hard to advise something without your code.
Could you please provide your sources?

**Comment by radumardale** [comment]
You are right! 
Here goes:

I have my AuthHttp that extends Http:

```
import {Injectable} from '@angular/core';
import {Http, Headers, Request, RequestOptionsArgs, Response} from '@angular/http';
import {Observable} from 'rxjs/Observable';

@Injectable()
export class AuthHttp extends Http {

    private token:string = "my-token";

    createAuthorizationHeader(headers:Headers) {
        headers.append('Authorization', 'Basic ' + this.token);
    }

    get(url:string, options?:RequestOptionsArgs):Observable<Response> {
        let headers = new Headers();
        this.createAuthorizationHeader(headers);
        return super.get(url, {
            headers: headers
        });
    }
}
```

I have an SecuredResource that extends Resource:

```
import {Resource, ResourceParams} from "ng2-resource-rest";
import {Http} from "@angular/http";
import {AuthHttp} from "./securedHttp";
@ResourceParams({
     add2Provides: false
})
export class SecResource extends Resource {
    constructor(protected http: Http) {
        super(<AuthHttp>http);
    }
}

```

And my UserResource

```
import {ResourceParams, ResourceAction, ResourceResult} from "ng2-resource-rest";
import {SecuredResource} from "../securedHttp/securedResource";

import {RequestMethod, Http} from "@angular/http";
import {Injectable} from "@angular/core";
import {HttpClient} from "../securedHttp/securedHttp";

@Injectable()
@ResourceParams({
    url: 'http://localhost:3030/api',
    path: '/v1/user/{id}'
})
export class UserResource extends SecuredResource {
    @ResourceAction({
        method: RequestMethod.Post,
        path: '/public/login'
    })
    login(data:any, callback?:Function):ResourceResult {
        return null;
    }
}

```

When using the UserResource in a component the call headers don't get sent:

```
        this.userResource.query({})
            .$observable
            .subscribe(
                res => {
                    console.log(res);
                },
                err => console.log('Err', err)
            );

```

I am trying to do this for authentication purpose. 

I have tried to extend Resource but I wasn't able to inject a service in there in order to get the auth token from it - in the example you posted with GCResource lastToken is static and it doesn't really help me. 

Basically I am trying to secure all my resources with a token from a service.

I have used ngResource with angular 1 a lot and it was really helpful but I have just started with angular 2 and things don't always go according to plan :)

Thank you.

**Comment by troyanskiy** [comment]
@radumardale 
1. Did you register you AuthHttp in bootstrap?
Could you please also provide your bootstrap?
2. I think it should be like that 
``` ts
import {Resource, ResourceParams} from "ng2-resource-rest";
import {Http} from "@angular/http";
import {AuthHttp} from "./securedHttp";
@ResourceParams({
     add2Provides: false
})
export class SecResource extends Resource {
    constructor(http: AuthHttp) {
        super(<Http>http);
    }
}
```

**Comment by radumardale** [comment]
1. I wasn't registering AuthHttp in bootstrap. 
I added it and nothing changed.

```
bootstrap(AppComponent, [
  disableDeprecatedForms(),
  provideForms(),
  HTTP_PROVIDERS,
  APP_ROUTER_PROVIDERS,
  RESOURCE_PROVIDERS,
  {
    provide: APP_BASE_HREF,
    useValue: '<%= APP_BASE %>'
  },
  SecuredHttp
]);

```

2. Tried this way also with no luck.

Can you think of any other way for implementing token based authentication?

**Comment by troyanskiy** [comment]
I'n my own project I'm implementing somthing like that
``` ts
import { Resource, ResourceParams } from 'ng2-resource-rest';

@ResourceParams({
  add2Provides: false,
})
export class BaseResource extends Resource {

  getHeaders():any {
    let headers = super.getHeaders();

    let token = localStorage.get('token');
    if (token) {
      headers['Authorization'] = 'Bearer ' + token;
    }

    return headers;
  }

}

```

**Comment by radumardale** [comment]
I have tried that but I was getting errors when trying to access localStorage in BaseResource. 

Are you using angular2-local-storage?

I will give this approach another try.

Thank you.

**Comment by troyanskiy** [comment]
I have AuthHelper class which is static class. It handle storing and retrieving tokens.
``` ts
import {LocalStorageHelper} from './localStorage.helper';
import {JwtHelper} from 'angular2-jwt';

export class AuthServiceHelper {
  static jwtHelper = new JwtHelper();
  private static _token:string = LocalStorageHelper.get('authToken');
  private static _tokenBody:any = null;

  static init() {
    if (this._token && this.jwtHelper.isTokenExpired(this._token)) {
      this.token = null;
    }
  }

  static extendHeaders(headers:any = {}):any {
    if (this._token) {
      headers['Authorization'] = 'Bearer ' + this._token;
    }
    return headers;
  }

  static get token() {
    return this._token;
  }

  static set token(newToken:string) {

    if (!newToken) {

      this._token = null;
      this._tokenBody = null;

    } else {

      this._token = newToken;
      this._tokenBody = AuthServiceHelper.jwtHelper.decodeToken(this._token);

    }

    LocalStorageHelper.set('authToken', this._token);
    LocalStorageHelper.set('authTokenBody', this._tokenBody);

  }

  static get tokenBody():any {
    return this._tokenBody;
  }

  static set tokenBody(body:any) {
    console.error('Can\'t set token body');
  }

}

```

`localStorage.helper.ts`
``` ts
export class LocalStorageHelper {


  static get(key: string):any {

    let value = localStorage.getItem(key);
    if (!value) {
      return null;
    }

    value = this.decode(value);
    if (this.isValid(value)) {
      return value.d;
    }

    localStorage.removeItem(key);

    return null;
  }

  static set(key: string, value: any, ttl: number = 0) {

    let obj:ICacheRecord = {
      v: ttl ? new Date().getTime() + ttl : 0,
      d: value
    };

    localStorage.setItem(key, this.encode(obj));

  }

  static remove(key: string) {
    localStorage.removeItem(key);
  }

  static removeAll() {
    localStorage.clear();
  }

  private static encode(value: ICacheRecord):string {
    return JSON.stringify(value);
  }

  private static decode(value: string):ICacheRecord {
    if (!value) {
      return null;
    }
    return JSON.parse(value);
  }

  private static isValid(value: ICacheRecord) {
    return !value.v || value.v > new Date().getTime();
  }




}

interface ICacheRecord {
  v: number;
  d: any;
}

```

**Comment by radumardale** [comment]
Thank you! 
This will do :)
